nemo/collections/nlp/models/nlp_model.py

```diff
@@ -410,3 +522,160 @@ def save_checkpoint(self, filepath, weights_only: bool):
+
+        if app_state.model_parallel_size is not None:
+
+            if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+
+                if app_state.model_parallel_group is None:
+                    self.init_model_parallel(app_state.global_rank, app_state.world_size)
+
+    def start_training(self, trainer: 'Trainer') -> None:
+        """ PTL Hook that is called after DPP is initialized. """
+
+        if isinstance(self.lightning_module.bert_model, MegatronBertEncoder):
+            app_state = AppState()
+            if app_state.model_parallel_size is not None:
+                # mpu grad clipping needs parameters to have the attribute model_parallel
+                parameters = self.lightning_module.parameters()
+                for p in parameters:
+                    if not hasattr(p, 'model_parallel'):
+                        p.model_parallel = False
+
+                # TODO: figure out how to override clip gradients again
+                # Update PTL trainer to use our _clip_gradients
+                # self._trainer.accelerator_backend._clip_gradients = self._clip_gradients
+
+                if get_checkpoint_version():
+                    # Restored from .nemo, checkpoint_version will already be set
+                    pass
+                elif trainer.resume_from_checkpoint is not None:
+                    # PTL auto-resuming, need to update checkpoint name
+                    # update path based on model parallel rank
+                    filepath = trainer.resume_from_checkpoint
+                    dirname = os.path.dirname(os.path.dirname(filepath))
+                    basename = os.path.basename(filepath)
+                    filepath = f'{dirname}/mp_rank_{app_state.model_parallel_rank:02d}/{basename}'
+                    trainer.resume_from_checkpoint = filepath
+                    logging.info(f'Resuming training from checkpoint {trainer.resume_from_checkpoint}')
+                    # need to set checkpoint version for megatron-lm
+                    checkpoint_version = torch.load(trainer.resume_from_checkpoint).get('checkpoint_version', None)
+                    if checkpoint_version is not None:
+                        set_checkpoint_version(checkpoint_version)
+                    else:
+                        logging.warning('Megatron-lm checkpoint version not found. Setting checkpoint_version to 0.')
+                        set_checkpoint_version(0)
+                else:
+                    logging.info(
+                        f"Restoring from pretrained model parallel checkpoint: {self.lightning_module.bert_model._restore_path}"
+                    )
+                    self.lightning_module.bert_model.restore_weights(self.lightning_module.bert_model._restore_path)
+
+            self.lightning_module.register_megatron_checkpoint_version()
+
+        return super().start_training(trainer)
+
+    def start_testing(self, trainer: 'Trainer') -> None:
```

Comment:
Just a question, theres no `start_validation`?

Reply:
That's correct.